### PR TITLE
bug/SWTCH-1250 Switchboard doesn't relogin on changing Metamask account

### PR DESCRIPTION
**Pull request by JGiter**
I can see two ways to enable signer service to listen on events emitted by `Metamask` or `WalletConnect` wallets:
- Underlying signer can extend `ethers.Signer` with `on` method
- wallet provider notifies signer service

I have chosen second way, because extending signer requires to implement `ethers.Signer` instead of getting it from `Web3Provider` wrapper from `ethers`. May be there is a better design

### CODE REVIEW

File: src/modules/signer/signer.types.ts [hunk 1]
```diff
@@ -8,9 +8,9 @@ export enum ProviderType {
 }
 
 export enum ProviderEvent {
```
Comment (jrhender) [inline src/modules/signer/signer.types.ts:10]: I guess these are from the wallet-connect or the metamask code? Are they documented somewhere? If so, could we add a link to where they are documented or defined?

File: src/modules/signer/signer.service.ts [hunk 2]
```diff
@@ -53,15 +54,18 @@ export class SignerService {
         this._servicesInitializers.push(initializer);
     }
 
-    on(event: ProviderEvent, cb) {
-        const provider = this._signer.provider;
-        if (provider instanceof EventEmitter) {
-            provider.on(event, cb);
-        } else if (provider instanceof WalletConnectProvider) {
-            provider.on(event, cb);
+    async emit(e: ProviderEvent) {
```
Comment (jrhender) [inline src/modules/signer/signer.service.ts:57]: I'm a bit confused by the naming here. It seems like this more of a `handleEvent` function 🤔 ? It doesn't look like anything is actually being emitted?

File: src/modules/signer/walletConnectMetamask.ts [hunk 3]
```diff
@@ -12,6 +12,8 @@ export const fromWalletConnectMetamask = async (bridge: string, infuraId?: strin
     await walletProvider.enable();
     const provider = new providers.Web3Provider(walletProvider);
     const signerService = new SignerService(provider.getSigner(), ProviderType.MetaMask);
```
Comment (jrhender) [inline src/modules/signer/walletConnectMetamask.ts:14]: This code seems weird. The provider is coming from `createWalletConnectProvider` but `SignerService` is initialized with `ProviderType.MetaMask`? Is this correct?

**Review by jrhender — COMMENTED** [review]
Just a few comments.

File: src/modules/signer/signer.service.ts (hunk 2 above)
Comment (JGiter) [reply src/modules/signer/signer.service.ts:57]: It is emitted in a sense that it notifies registered listeners. It is similar to Node.js EventEmiter.emit

File: src/modules/signer/walletConnectMetamask.ts (hunk 3 above)
Comment (JGiter) [reply src/modules/signer/walletConnectMetamask.ts:14]: thanks for noticing. may be this ProviderType isn't needed really. it is only used to register listeners of connected provider, but it wouldn't harm if listeners of other providers was registered

Comment (jrhender) [reply src/modules/signer/walletConnectMetamask.ts:14]: > it is only used to register listeners of connected provider
I don't understand. What is the relationship of this class to WalletConnect or Metamask?

Comment (JGiter) [reply src/modules/signer/walletConnectMetamask.ts]: This class is notified by wallet to clean session or reinitialize itself when account is changed
 https://github.com/energywebfoundation/iam-client-lib/blob/44e0ad3ecb8c3fecc98bd717e389bbb60cfbbe2e/src/modules/signer/signer.service.ts#L78

File: src/modules/signer/signer.service.ts [hunk 4]
```diff
@@ -71,10 +75,13 @@ export class SignerService {
...
-        this.on(ProviderEvent.AccountChanged, accChangedHandler);
-        this.on(ProviderEvent.NetworkChanged, accChangedHandler);
-        this.on(ProviderEvent.Disconnected, this.closeConnection());
-        this.on(ProviderEvent.SessionUpdate, accChangedHandler);
+        if (this._providerType === ProviderType.MetaMask) {
+            this.on(ProviderEvent.AccountChanged, accChangedHandler);
+            this.on(ProviderEvent.NetworkChanged, accChangedHandler);
+        } else if (this._providerType === ProviderType.WalletConnect) {
+            this.on(ProviderEvent.SessionUpdate, accChangedHandler);
+            this.on(ProviderEvent.Disconnected, this.closeConnection());
```
Comment (Harasz) [inline src/modules/signer/signer.service.ts:83]: Shouldn't be like this?
```suggestion
            this.on(ProviderEvent.Disconnected, this.closeConnection);
```

File: src/modules/signer/signer.service.ts [hunk 5]
```diff
@@ -53,15 +54,18 @@ export class SignerService {
...
     }
 
-    on(event: ProviderEvent, cb) {
-        const provider = this._signer.provider;
-        if (provider instanceof EventEmitter) {
-            provider.on(event, cb);
-        } else if (provider instanceof WalletConnectProvider) {
-            provider.on(event, cb);
+    async emit(e: ProviderEvent) {
+        for await (const { event, cb } of this._walletEventListeners) {
```
Comment (Harasz) [inline src/modules/signer/signer.service.ts:58]: Could we wrap it to `Promise.all` to increase performance?

File: src/modules/signer/walletConnectMetamask.ts [hunk 6]
```diff
@@ -11,7 +11,9 @@ export const fromWalletConnectMetamask = async (bridge: string, infuraId?: strin
     const walletProvider = createWalletConnectProvider(bridge, infuraId);
     await walletProvider.enable();
     const provider = new providers.Web3Provider(walletProvider);
-    const signerService = new SignerService(provider.getSigner(), ProviderType.MetaMask);
+    const signerService = new SignerService(provider.getSigner(), ProviderType.WalletConnect);
+    walletProvider.on(ProviderEvent.Disconnected, signerService.emit(ProviderEvent.Disconnected));
```
Comment (Harasz) [inline src/modules/signer/walletConnectMetamask.ts:15]:
```suggestion
    walletProvider.on(ProviderEvent.Disconnected, () => signerService.emit(ProviderEvent.Disconnected));
```

File: src/modules/signer/walletConnectMetamask.ts [hunk 7]
```diff
@@ -11,7 +11,9 @@ export const fromWalletConnectMetamask = async (bridge: string, infuraId?: strin
...
+    walletProvider.on(ProviderEvent.SessionUpdate, signerService.emit(ProviderEvent.SessionUpdate));
```
Comment (Harasz) [inline src/modules/signer/walletConnectMetamask.ts:16]:
```suggestion
    walletProvider.on(ProviderEvent.SessionUpdate, () => signerService.emit(ProviderEvent.SessionUpdate));
```

**Review by Harasz — CHANGES_REQUESTED** [review]

**Review by Harasz — APPROVED** [review]

**Review by jrhender — APPROVED** [review]
Looks good to me!

**Pull request merged**